src/hotspot/share/ci/ciMemberName.cpp

```diff
@@ -34,11 +34,16 @@
 // Return: MN.vmtarget
 ciMethod* ciMemberName::get_vmtarget() const {
   VM_ENTRY_MARK;
-  // FIXME: Share code with ciMethodHandle::get_vmtarget
-  Metadata* vmtarget = java_lang_invoke_MemberName::vmtarget(get_oop());
-  if (vmtarget->is_method())
+  return get_vmtarget_method(get_oop());
+}
+
+ciMethod* ciMemberName::get_vmtarget_method(oop mname) {
+  ASSERT_IN_VM;
+  Metadata* vmtarget = java_lang_invoke_MemberName::vmtarget(mname);
+  if (vmtarget->is_method()) {
     return CURRENT_ENV->get_method((Method*) vmtarget);
-  // FIXME: What if the vmtarget is a Klass?
-  assert(false, "");
-  return NULL;
+  } else {
+    fatal("vmtarget should be a method");
```

Comment:
There's a slight difference in behavior here: `fatal()` is present in product binaries while `assert()` is not. 
Is the change deliberate?

Reply:
Yes. Because AFAICS, the `get_vmtarget` callers use the result without any null checks, calling member methods off that `ciMethod*`, and thus release bits would (hopefully) `SEGV` if this path is taken. In this case, it seems prudent to `fatal()` at sensible point before that happens.

Reply:
Ok. I took a closer look and noticed that `java_lang_invoke_MemberName::vmtarget` already returns `Method*`. So, the code can be rewritten as follows:
```
  Method* vmtarget = java_lang_invoke_MemberName::vmtarget(get_oop());
  return CURRENT_ENV->get_method(vmtarget);
```

Reply:
Oh, good stuff. I missed that! This means we can simplify the whole patch, and avoid the new method and the dependency between `ciMethodHandle` and `ciMemberName`. Please see new patch.